File: src/main/java/net/hectus/neobb/game/mode/CardGame.java
package net.hectus.neobb.game.mode;

import com.marcpg.libpg.data.time.Time;
import com.marcpg.libpg.lang.Translation;
import net.hectus.neobb.NeoBB;
import net.hectus.neobb.game.Game;
import net.hectus.neobb.game.util.GameInfo;
import net.hectus.neobb.lore.CardItemLoreBuilder;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.shop.RandomizedShop;
import net.hectus.neobb.turn.Turn;
import net.hectus.neobb.turn.card_game.*;
import net.hectus.neobb.turn.default_game.warp.TDefaultWarp;
import net.hectus.neobb.util.Colors;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Locale;

/**
 * Source: <a href="https://github.com/TiagoFar78/BlockBattles">Source Code on GitHub</a>
 * <p>Originally named HereGame, because it was based on HereStudio, but renamed to CardGame.
 */
public class CardGame extends Game {
    public static final GameInfo INFO = new GameInfo(true, false, 50.0, 0, 5, new Time(10, Time.Unit.MINUTES), 15, RandomizedShop.class, CardItemLoreBuilder.class, List.of(
            CTChest.class, CTDaylightDetector.class, CTFlowerPot.class, CTJackOLantern.class, CTOakDoor.class, CTOakFenceGate.class,
            CTPointedDripstone.class, CTRedstoneLamp.class, CTTorch.class, CTWaxedExposedCutCopperStairs.class
    ));

    public CardGame(boolean ranked, World world, @NotNull List<Player> players) {
        super(ranked, world, players, new TDefaultWarp(world));
    }

    @Override
    public GameInfo info() {
        return INFO;
    }

    @Override
    public void postTurn(@NotNull Turn<?> turn, boolean skipped) {
        turn.player().inventory.fillInRandomly();
    }

    @Override
    public @Nullable List<Component> scoreboard(@NotNull NeoPlayer player) {
        Locale l = player.locale();
        try {
            return List.of(
                    MiniMessage.miniMessage().deserialize("<gradient:#D068FF:#EC1A3D>BlockBattles<reset><#BF646B>-<#9D9D9D>Alpha"),
                    Translation.component(l, "scoreboard.turning").color(Colors.BLUE)
                            .append(Component.text(currentPlayer() == player ? "You" : currentPlayer().player.getName(), Colors.RESET)),
                    Translation.component(l, "scoreboard.time").color(Colors.BLUE)
                            .append(Component.text(player.game.timeLeft().getPreciselyFormatted(), Colors.RESET)),
                    Translation.component(l, "scoreboard.health").color(Colors.BLUE)
                            .append(Component.text(player.health(), Colors.RESET)),
                    Component.empty(),
                    Component.text("NeoBB-" + NeoBB.VERSION + " (d" + Integer.toHexString(LocalDateTime.now().getDayOfYear()) + "h" + LocalDateTime.now().getHour() + ")", Colors.EXTRA),
                    Component.text("mc.hectus.net", Colors.LINK)
            );
        } catch (Exception e) {
            return List.of(MiniMessage.miniMessage().deserialize("<gradient:#D068FF:#EC1A3D>BlockBattles<reset><#BF646B>-<#9D9D9D>Alpha"));
        }
    }

    @Override
    public boolean allows(Turn<?> turn) {
        return true;
    }
}

File: src/main/kotlin/net/hectus/neobb/game/mode/CardGame.kt
package net.hectus.neobb.game.mode

import com.marcpg.libpg.display.*
import com.marcpg.libpg.lang.string
import com.marcpg.libpg.util.component
import net.hectus.neobb.NeoBB
import net.hectus.neobb.game.Game
import net.hectus.neobb.game.util.GameDifficulty
import net.hectus.neobb.game.util.GameInfo
import net.hectus.neobb.modes.lore.CardItemLoreBuilder
import net.hectus.neobb.modes.shop.RandomizedShop
import net.hectus.neobb.modes.turn.Turn
import net.hectus.neobb.modes.turn.TurnExec
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.util.Colors
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.World
import org.bukkit.entity.Player
import java.time.LocalDateTime

/**
 * Source: [Source Code on GitHub](https://github.com/TiagoFar78/BlockBattles)
 *
 * Originally named HereGame, because it was based on HereStudio, but renamed to CardGame.
 */
class CardGame(world: World, bukkitPlayers: List<Player>, difficulty: GameDifficulty = GameDifficulty.NORMAL) : Game(world, bukkitPlayers, difficulty) {
    override val info: GameInfo = GameInfo(
        namespace = "card",
        hasStructures = false,
        startingHealth = 40.0,
        coins = 0,
        deckSize = 5,
        turnTimer = 15,
        shop = RandomizedShop::class,
        loreBuilder = CardItemLoreBuilder::class
    )

    override val scoreboard: ((NeoPlayer) -> SimpleScoreboard)? = { p -> SimpleScoreboard(p, 5, MiniMessage.miniMessage().deserialize("<gradient:#D068FF:#EC1A3D>BlockBattles<reset><#BF646B>-<#9D9D9D>Alpha"),
        ValueScoreboardEntry(p.locale().component("scoreboard.turning", color = Colors.BLUE)) { component(if (currentPlayer() === p) p.locale().string("scoreboard.turning.you") else currentPlayer().name()) },
        ValueScoreboardEntry(p.locale().component("scoreboard.time", color = Colors.BLUE)) { component(p.game.timeLeft.preciselyFormatted) },
        ValueScoreboardEntry(p.locale().component("scoreboard.health", color = Colors.BLUE)) { Component.text(p.health) },
        BlankScoreboardEntry(),
        SimpleScoreboardEntry { component("NeoBB-" + NeoBB.VERSION + " (d" + Integer.toHexString(LocalDateTime.now().dayOfYear) + "h" + LocalDateTime.now().hour + ")", Colors.EXTRA) },
        StaticScoreboardEntry(component("mc.hectus.net", Colors.LINK)),
    ) }

    override fun postTurn(exec: TurnExec<*>, skipped: Boolean) {
        exec.player.inventory.fillInRandomly()
    }

    // Always return true, basically disabling any usage rules:
    override fun allows(turn: Turn<*>): Boolean = true
}
